### PR TITLE
Fix issues with records and support instanceof final

### PR DESCRIPTION
**Pull request by bkramer-relyance**
This PR continues the work of supporting Java records, letting us:
1. define a record that’s nested in an interface, not just a class.
2. inherit from an interface
3. define a compact constructor 

See https://docs.oracle.com/en/java/javase/18/language/records.html.

Caveat: to reduce the complexity of the grammar/parser, I defined compact constructors to be accepted at any place one might find a normal constructor, making the tree-sitter parser more permissive than the spec. I'm going by the assumption that the philosophy of tree-sitter is to understand the correct code, not to enforce it. (I could have made the `compact_constructor_declaration` rule by allowing methods to have no arguments, but felt that was taking it too far.) I'm open to some guidance on this topic.

Because it was convenient for me (fewer versions of generated code to keep track of), I am also adding in this PR support for the optional `final` modifier after `instanceof`. I can't find the spec for this, but here's some communication concerning it: https://github.com/checkstyle/checkstyle/issues/9897

All of these issues were discovered in the Elasticsearch repo.

Checklist:
- [x] All tests pass in CI.
- [x] There are sufficient tests for the new fix/feature.
- [x] Grammar rules have not been renamed unless absolutely necessary.
- [x] The conflicts section hasn't grown too much.
- [x] The parser size hasn't grown too much (check the value of STATE_COUNT in src/parser.c).

Question: while I was working out the grammar updates for records, I noticed that there is the inclusion of `record` as a "reserved identifier," along with just a couple of others.  This seems pretty arbitrary and I'm wondering if they are to deal with cases where the Java language permits these keywords to be repurposed as identifiers, like with lambda arguments. I've seen this in one of the corpus tests, as well as for `module` in one of the Github repos we use to test. I also see a current parse failure in the Ghidra repo where `sealed` appears as a variable name undergoing assignment.  So what's the general story here? Are we just opportunistically adding reserved identifiers to deal with these discoveries, and if so, isn't the naming backward (we are not "reserving keywords" but rather "un-reserving them.")?

### CODE REVIEW

**Comment by bkramer-relyance** [comment]
Changes in parser.c counts:
Before:
```
#define LANGUAGE_VERSION 14
#define STATE_COUNT 1267
#define LARGE_STATE_COUNT 350
#define SYMBOL_COUNT 291
#define ALIAS_COUNT 1
#define TOKEN_COUNT 130
#define EXTERNAL_TOKEN_COUNT 0
#define FIELD_COUNT 37
#define MAX_ALIAS_SEQUENCE_LENGTH 11
#define PRODUCTION_ID_COUNT 231
```
After:
```
#define LANGUAGE_VERSION 14
#define STATE_COUNT 1282
#define LARGE_STATE_COUNT 354
#define SYMBOL_COUNT 292
#define ALIAS_COUNT 1
#define TOKEN_COUNT 130
#define EXTERNAL_TOKEN_COUNT 0
#define FIELD_COUNT 37
#define MAX_ALIAS_SEQUENCE_LENGTH 11
#define PRODUCTION_ID_COUNT 238
```

**Comment by bkramer-relyance** [comment]
Awaiting review and CI testing.

**Comment by bkramer-relyance** [comment]
@aryx Requesting review.

**Comment by aryx** [comment]
@bmahe is this related to your recent PR?

**Review by aryx — APPROVED** [review]
LGTM.
@bmahe ?

**Comment by aryx** [comment]
Could you rebase to solve the conflicts?
I'll merge after that.

File: grammar.js
Comment (bkramer-relyance) [reply grammar.js]: Fixed.

**Comment by bkramer-relyance** [comment]
@aryx I rebased my branch on master and force pushed.  This could use another round of testing before merging. Thank you.

**Review by aryx — APPROVED** [review]

**Pull request merged**
